alacritty_terminal/src/config/mod.rs

```diff
@@ -418,17 +401,50 @@ impl Default for DefaultTrueBool {
     }
 }
 
+fn defaulted<T, E>(err: E) -> T
+    where T: Default, E: Display
+{
+    error!("Problem with config: {}; using default value", err);
+    T::default()
+}
+
 pub fn failure_default<'a, D, T>(deserializer: D) -> Result<T, D::Error>
 where
     D: Deserializer<'a>,
     T: Deserialize<'a> + Default,
 {
-    let value = serde_yaml::Value::deserialize(deserializer)?;
-    match T::deserialize(value) {
-        Ok(value) => Ok(value),
-        Err(err) => {
-            error!("Problem with config: {}; using default value", err);
-            Ok(T::default())
-        },
-    }
+    Ok(T::deserialize(Value::deserialize(deserializer)?).unwrap_or_else(defaulted))
+}
+
+pub fn none_str_as_none<'de, T, D>(deserializer: D) -> Result<Option<T>, D::Error>
+    where D: Deserializer<'de>, T: Deserialize<'de> + Default
+{
+    Ok(match Value::deserialize(deserializer)? {
+        Value::String(ref value) if value.to_lowercase() == "none" => None,
+        value => Some(T::deserialize(value).unwrap_or_else(defaulted))
+    })
+}
+
+pub fn abbrev_from_str<'de, T, D>(deserializer: D) -> Result<T, D::Error>
```

Comment:
I think the naming of this is also a bit troublesome. I wouldn't know what `abbrev_from_str` does at all for example. Maybe `deserialize_string_or_generic` might be better? The `or_generic` part is kinda ugly, but I can't think of anything better myself right now.

Reply:
I could go back to `struct_or_str`, which is more likely to be recognized since it's the name used in the official serde documentation (although that version enforces the requirement that it be a struct if it's not a string).

Reply:
Generally `str` is not a good abbreviation here, since it's easily confused with the Rust type. Which is probably why the serde documentation example also uses `string_or_struct`. I believe the yaml type is also called `String`.

Generally for one-off deserializers, I've also added the `deserialize_` at the beginning to make clear what they are for. `failure_default` is an exception here simply because it's used everywhere and previously there was an additional function which has been relplaced with `#[serde(default)]` which is why a shorter name was beneficial.

Having `struct` in the name is also misleading. This is in no way limited to structs, so the function name would actually be completely incorrect. We either deserialize from a serde string, or from whatever other type the field has which implements deserialize. This is why I proposed `or_generic` instead of `or_struct`.

Generally another problem that could be mentioned is that `string` does not explicitly mention that its about the origin and not the target type (neither in my example nor yours). So `deserialize_string_or_generic` could be easily mistaken for a deserializer which deserializes either a field with the value string, or a generic value.

I'll happily admit that I'm not really happy with any solution proposed so far, neither by you nor me. I'm not sure if there is a perfect solution either, so the better choice might be to just settle for something reasonable.

Considering most of the points I've made, maybe `from_string_or_deserialize` could be an option? It's a quite literal way to translate what exactly is happening, so it might be a relatively uncontroversial solution that still represents things accurately. It doesn't really say that from string is also deserialization, but I guess you can't win them all.

Reply:
`from_string_or_deserialize` sounds pretty good